test/jdk/java/util/IdentityHashMap/Basic.java

```diff
@@ -0,0 +1,546 @@
+        assertEquals(map.size(), 3);
+
+        assertTrue(map.containsKey(k1a));
+        assertTrue(map.containsKey(k1b));
+        assertTrue(map.containsKey(k2));
+        assertFalse(map.containsKey(new Box(k1a)));
+
+        assertTrue(map.containsValue(v1a));
+        assertTrue(map.containsValue(v1b));
+        assertFalse(map.containsValue(new Box(v1a)));
+        assertTrue(map.containsValue(v2));
+    }
+
+    // get
+    @Test
+    public void testGet() {
+        assertSame(map.get(k1a), v1a);
+        assertSame(map.get(k1b), v1b);
+        assertSame(map.get(k2), v2);
+        assertNull(map.get(new Box(k1a)));
+    }
+
+    // getOrDefault
+    @Test
+    public void testGetOrDefault() {
+        Box other = new Box(22);
+
+        assertSame(map.getOrDefault(k1a, other), v1a);
+        assertSame(map.getOrDefault(k1b, other), v1b);
+        assertSame(map.getOrDefault(new Box(k1a), other), other);
+        assertSame(map.getOrDefault(k2, other), v2);
+    }
+
+    // clear
+    // isEmpty
+    @Test
+    public void testClearEmpty() {
+        assertFalse(map.isEmpty());
+        map.clear();
+        assertTrue(map.isEmpty());
+    }
+
+    // hashCode
+    @Test
+    public void testHashCode() {
+        int expected = (System.identityHashCode(k1a) ^ System.identityHashCode(v1a)) +
+                       (System.identityHashCode(k1b) ^ System.identityHashCode(v1b)) +
+                       (System.identityHashCode(k2)  ^ System.identityHashCode(v2));
+        assertEquals(map.hashCode(), expected);
+        assertEquals(map.entrySet().hashCode(), expected);
+    }
+
+    // equals
+    @Test
+    public void testEquals() {
+        assertTrue(map.equals(map));
+        assertTrue(map.equals(map2));
+        assertTrue(map2.equals(map));
+    }
+
+    // equals
+    @Test
+    public void testEqualsDifferentKey() {
+        map2.remove(k1a);
+        map2.put(new Box(k1a), v1a);
+        assertFalse(map.equals(map2));
+        assertFalse(map2.equals(map));
+    }
+
+    // equals
+    @Test
+    public void testEqualsDifferentValue() {
+        map2.put(k1a, new Box(v1a));
+        assertFalse(map.equals(map2));
+        assertFalse(map2.equals(map));
+    }
+
+    @Test
+    public void testKeySet() {
+        Set<Box> keySet = map.keySet();
+        checkElements(keySet, k1a, k1b, k2);
+        assertFalse(keySet.contains(new Box(k1a)));
+    }
+
+    @Test
+    public void testKeySetNoRemove() {
+        Set<Box> keySet = map.keySet();
+        keySet.remove(new Box(k1a));
+        checkElements(keySet, k1a, k1b, k2);
+        assertTrue(keySet.equals(map2.keySet()));
+        assertTrue(map.equals(map2));
+    }
+
+    @Test
+    public void testKeySetRemove() {
+        Set<Box> keySet = map.keySet();
+        keySet.remove(k1a);
+        checkElements(keySet, k1b, k2);
+        checkEntries(map.entrySet(), entry(k1b, v1b),
+                                     entry(k2, v2));
+    }
```

Comment:
Would an additional check `assertFalse(map.equals(map2));` be useful here (and other similar tests where we do "remove").

Reply:
I don't know if you noticed that previous versions checked the map's contents with `map.equals(map2)` and either asserted true or false depending on the situation. I pulled most of those out when I added `checkEntries`. The reason is that `checkEntries` and `checkElements` are supposed to check the exact contents of the map or the collection, and they fail if there is a missing, different, or extra entry or element. If that's true we don't need to test `map.equals`. I don't think it calling `map.equals` adds any value or does any checking that the `checkX` methods don't already do.

Of course this relies on `checkEntries` and `checkElements` to do their jobs properly, so we should make sure they do!

And also we need to test that the `equals` method is doing its job as well. There are a couple tests for it already, and they test at least the basic cases. But it's possible I might have missed something.

In any case, if we believe the `checkX` methods are sufficient, and if we believe that the tests for `equals` are also sufficient, then I don't think we need to add assertions about `equals` in any tests other than for `equals` itself.